enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/BoltInteraction.java

```diff
@@ -0,0 +1,318 @@
+            fieldNames = (List<String>) metadata.get( "fields" );
+        }
+        else if ( message instanceof FailureMessage )
+        {
+            FailureMessage failMessage = ((FailureMessage) message);
+            client.send( TransportTestUtil.chunk( reset() ) );
+            collectResults( client );
+            throw new AuthenticationException( failMessage.status(), failMessage.message() );
+        }
+
+        do
+        {
+            message = TransportTestUtil.recvOneMessage( client );
+            if ( message instanceof RecordMessage )
+            {
+                Object[] row = ((RecordMessage) message).record().fields();
+                Map<String,Object> rowMap = new HashMap<>();
+                for ( int i = 0; i < row.length; i++ )
+                {
+                    rowMap.put( fieldNames.get( i ), row[i] );
+                }
+                result.add( rowMap );
+            }
+        }
+        while ( !(message instanceof SuccessMessage) && !(message instanceof FailureMessage) );
+
+        if ( message instanceof FailureMessage )
+        {
+            FailureMessage failMessage = ((FailureMessage) message);
+            client.send( TransportTestUtil.chunk( reset() ) );
+            TransportTestUtil.recvOneMessage( client );
+            throw new AuthenticationException( failMessage.status(), failMessage.message() );
+        }
+
+        return new BoltResult( result );
+    }
+
+    public static class BoltSubject
+    {
+        Connection client;
+        String username;
+        String password;
+        AuthenticationResult loginResult = AuthenticationResult.FAILURE;
+
+        public BoltSubject( Connection client, String username, String password )
+        {
+            this.client = client;
+            this.username = username;
+            this.password = password;
+        }
+
+        public void setLoginResult( Message result )
+        {
+            if ( result instanceof SuccessMessage )
+            {
+                Map<String,Object> meta = ((SuccessMessage) result).meta();
+                if ( meta.containsKey( "credentials_expired" ) && meta.get( "credentials_expired" ).equals( true ) )
+                {
+                    loginResult = AuthenticationResult.PASSWORD_CHANGE_REQUIRED;
+                }
+                else
+                {
+                    loginResult = AuthenticationResult.SUCCESS;
+                }
+            }
+            else if ( result instanceof FailureMessage )
+            {
+                Status status = ((FailureMessage) result).status();
+                if ( status.equals( Status.Security.Unauthorized ) )
+                {
+                    loginResult = AuthenticationResult.FAILURE;
+                }
+            }
+        }
+
+        public boolean isAuthenticated()
+        {
+            return loginResult.equals( AuthenticationResult.SUCCESS );
```

Comment:
Interesting, when in PASSWORD_CHANGE_REQUIRED, you're not authenticated? This is probably a convenient change.